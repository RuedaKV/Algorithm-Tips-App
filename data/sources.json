{
    "federal": [
        "Federal Agency - Executive",
        "Federal Agency - Judicial",
        "Federal Agency - Legislative"
    ],
    "regional": [
        "State/Local Govt",
        "Interstate Agency",
        "Native Sovereign Nation",
        "Independent Intrastate Agency"
    ],
    "local": [
        "City",
        "County"
    ]
}
