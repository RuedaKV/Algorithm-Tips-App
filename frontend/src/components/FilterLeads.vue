<template>
  <div>
    <div class="row">
      <b-form @submit.prevent="submitSearch" class="col-sm-12" id="lead-filter">
        <b-form-input v-model="form.filter" placeholder="Search..." />
        <div id="advanced-toggle" v-b-toggle:advanced-filters>
          <b-icon-chevron-right class="when-closed" />
          <b-icon-chevron-down class="when-open" />
          <span>Advanced Filters</span>
        </div>
        <b-collapse id="advanced-filters">
          <b-form-group
            label="Discovered Before:"
            label-for="to-date"
            label-cols-sm="3"
            label-align="right"
          >
            <b-form-datepicker id="to-date" v-model="form.to" />
          </b-form-group>
          <b-form-group
            label="Discovered After:"
            label-for="from-date"
            label-cols-sm="3"
            label-align="right"
          >
            <b-form-datepicker id="from-date" v-model="form.from" />
          </b-form-group>
        </b-collapse>

        <b-button-group id="submit-group" class="float-right">
          <b-button to="/db">Reset</b-button>
          <b-button type="submit" :to="search_path" variant="primary">Search</b-button>
        </b-button-group>
      </b-form>
    </div>
    <div class="row justify-content-center" v-if="loading">
      <b-spinner />
    </div>
    <div class="row justify-content-center" v-else-if="no_results">
      <p>
        <em>No results found</em>
      </p>
    </div>
    <div id="#leads" v-else>
      <div v-bind:key="lead.id" v-for="lead in leads" class="row justify-content-center">
        <Lead v-bind="lead" header-link />
      </div>
      <b-pagination-nav
        :number-of-pages="page_count"
        v-model="page"
        no-page-detect
        use-router
        :link-gen="linkGen"
      />
    </div>
  </div>
</template>

<script>
import Lead from "./Lead.vue";
import { mapActions, mapGetters } from "vuex";

export default {
  name: "FilterLeads",
  props: {
    query: Object
  },
  components: {
    Lead
  },
  data() {
    return {
      leads: null,
      form: {
        filter: this.query.filter,
        from: this.query.from,
        to: this.query.to,
        source: this.query.source
      },
      page_count: 1
    };
  },
  methods: {
    ...mapActions({
      filterLeads: "leads/filter"
    }),
    linkGen(pagenum) {
      return {
        path: "/db",
        query: { ...this.clean_filter(), page: pagenum }
      };
    },
    updateData() {
      this.leads = null;
      const filter = this.query;
      this.filterLeads({ params: filter, page: this.page }).then(() => {
        this.leads = this.getFilter(filter, this.page);
        this.page_count = this.getFilterPages(filter);
      });
    },
    clearForm() {
      this.filter = {
        filter: undefined,
        from: undefined,
        to: undefined,
        source: undefined
      };
    },
    submitSearch() {
      this.$router.push(this.search_path);
    },
    clean_filter() {
      return Object.fromEntries(
        Object.entries(this.form).filter(([, value]) => value !== undefined)
      );
    }
  },
  computed: {
    ...mapGetters({
      getFilter: "leads/filter-get",
      getFilterPages: "leads/filter-pages"
    }),
    page() {
      return this.query.page || 1;
    },
    loading() {
      return this.leads === null;
    },
    no_results() {
      return this.leads !== null && this.leads.length === 0;
    },
    search_path() {
      return {
        path: "/db",
        query: {
          ...this.clean_filter(),
          page: 1
        }
      };
    }
  },
  mounted() {
    this.updateData();
  }
};
</script>

<style lang="stylus" scoped>
#lead-filter {
  margin-bottom: 2em;
}

#submit-group {
  margin-top: 1em;
}

.collapsed > .when-open, .not-collapsed > .when-closed {
  display: none;
}

#advanced-toggle {
  margin-top: 0.5em;
}

#advanced-toggle > span {
  margin-left: 0.5em;
}
</style>