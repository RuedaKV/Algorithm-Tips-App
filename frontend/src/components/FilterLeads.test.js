import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import { BootstrapVue, IconsPlugin } from 'bootstrap-vue';
import Vue from 'vue';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import FilterLeads from './FilterLeads.vue';
import data from '../../test/test-lead.json';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);
localVue.use(BootstrapVue);
localVue.use(IconsPlugin);

describe('FilterLeads', () => {
    let store;
    let leads_module;
    let router;

    beforeEach(() => {
        router = new VueRouter();

        leads_module = {
            namespaced: true,
            actions: {
                load: jest.fn(() => Promise.resolve()),
                filter: jest.fn(() => Promise.resolve()),
            },
            getters: {
                find: jest.fn(),
                'filter-get': jest.fn(() => () => [data]),
                'filter-pages': jest.fn(() => () => 1),
            }
        };

        store = new Vuex.Store({
            modules: { leads: leads_module }
        });
    });

    const makeFilter = (fn) => (query) => fn(FilterLeads, {
        propsData: {
            query,
        },
        localVue,
        store,
        router,
    });

    const shallowFilter = makeFilter(shallowMount);
    const mountFilter = makeFilter((q, opts) => mount(q, {
        attachToDocument: true,
        ...opts
    }));

    it('should call filter on mount with the query param values', async () => {
        const query = {
            filter: 'test',
            from: '2020-1-20',
        };

        const el = shallowFilter(query);

        expect(leads_module.actions.filter.mock.calls[0][1]).toEqual({
            params: query,
            page: 1,
        });

        await Vue.nextTick();
        expect(el.find('lead-stub').props('name')).toBe(data.name);

        expect(el).toMatchSnapshot();
    });

    it('should set the query string after submitting the search form', async () => {
        const query = {};
        const el = mountFilter(query);

        el.find("#filter-input").setValue('test');
        await el.find("#lead-filter").trigger('submit');

        await Vue.nextTick();

        expect(router.currentRoute.query).toEqual({
            filter: 'test'
        });

        el.destroy();
    });

    it('should not search for an empty string', async () => {
        const query = {
            to: '2020-01-01',
        };
        const el = mountFilter(query);

        el.find("#filter-input").setValue('');
        await el.find("#lead-filter").trigger('submit');

        await Vue.nextTick();

        expect(router.currentRoute.query).toEqual({
            to: '2020-01-01'
        });

        el.destroy();

    });
});