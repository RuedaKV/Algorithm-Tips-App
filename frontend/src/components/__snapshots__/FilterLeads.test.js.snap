// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterLeads should call filter on mount with the query param values 1`] = `
<login-required-stub disabled="true">
  <div id="leads-container">
    <div class="row">
      <b-form-stub id="lead-filter" class="col-sm-12">
        <b-form-input-stub id="filter-input" value="test" placeholder="Search..." debounce="0" type="text"></b-form-input-stub>
        <div id="advanced-toggle" role="button" tabindex="0" class="collapsed" aria-expanded="false" aria-controls="advanced-filters">
          <b-icon-chevron-right-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-closed"></b-icon-chevron-right-stub>
          <b-icon-chevron-down-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-open"></b-icon-chevron-down-stub> <span>Advanced Filters</span>
        </div>
        <b-collapse-stub id="advanced-filters" tag="div">
          <b-form-group-stub label="Discovered On or After:" labelfor="from-date" labelcolssm="3" labelalign="right" feedbackarialive="assertive">
            <b-form-datepicker-stub id="from-date" value="2020-1-20" startweekday="0" buttonvariant="secondary" calendarwidth="270px" selectedvariant="primary" labeltodaybutton="Select today" todaybuttonvariant="outline-primary" labelresetbutton="Reset" resetbuttonvariant="outline-danger" labelclosebutton="Close" closebuttonvariant="outline-secondary" labelprevdecade="Previous decade" labelprevyear="Previous year" labelprevmonth="Previous month" labelcurrentmonth="Current month" labelnextmonth="Next month" labelnextyear="Next year" labelnextdecade="Next decade" labeltoday="Today" labelselected="Selected date" labelnodateselected="No date selected" labelcalendar="Calendar" labelnav="Calendar navigation" labelhelp="Use cursor keys to navigate calendar dates" dateformatoptions="[object Object]" weekdayheaderformat="short" offset="0" boundary="scrollParent"></b-form-datepicker-stub>
          </b-form-group-stub>
          <b-form-group-stub label="Discovered On or Before:" labelfor="to-date" labelcolssm="3" labelalign="right" feedbackarialive="assertive">
            <b-form-datepicker-stub id="to-date" startweekday="0" buttonvariant="secondary" calendarwidth="270px" selectedvariant="primary" labeltodaybutton="Select today" todaybuttonvariant="outline-primary" labelresetbutton="Reset" resetbuttonvariant="outline-danger" labelclosebutton="Close" closebuttonvariant="outline-secondary" labelprevdecade="Previous decade" labelprevyear="Previous year" labelprevmonth="Previous month" labelcurrentmonth="Current month" labelnextmonth="Next month" labelnextyear="Next year" labelnextdecade="Next decade" labeltoday="Today" labelselected="Selected date" labelnodateselected="No date selected" labelcalendar="Calendar" labelnav="Calendar navigation" labelhelp="Use cursor keys to navigate calendar dates" dateformatoptions="[object Object]" weekdayheaderformat="short" offset="0" boundary="scrollParent"></b-form-datepicker-stub>
          </b-form-group-stub>
          <b-form-group-stub label="Included Sources:" labelcolssm="3" labelalign="right" feedbackarialive="assertive">
            <div class="row ml-1">
              <div class="col-sm-4">
                <b-form-checkbox-stub value="true" checked="true">Federal</b-form-checkbox-stub>
                <b-form-select-stub options="[object Object],Federal Agency - Executive,Federal Agency - Judicial,Federal Agency - Legislative" valuefield="value" textfield="text" htmlfield="html" disabledfield="disabled" labelfield="label" optionsfield="options" selectsize="0"></b-form-select-stub>
              </div>
              <div class="col-sm-4">
                <b-form-checkbox-stub value="true" checked="true">State / Regional</b-form-checkbox-stub>
                <b-form-select-stub options="[object Object],State/Local Govt,Interstate Agency,Native Sovereign Nation" valuefield="value" textfield="text" htmlfield="html" disabledfield="disabled" labelfield="label" optionsfield="options" selectsize="0"></b-form-select-stub>
              </div>
              <div class="col-sm-4">
                <b-form-checkbox-stub value="true" checked="true">Local</b-form-checkbox-stub>
                <b-form-select-stub options="[object Object],City,County" valuefield="value" textfield="text" htmlfield="html" disabledfield="disabled" labelfield="label" optionsfield="options" selectsize="0"></b-form-select-stub>
              </div>
            </div>
          </b-form-group-stub>
        </b-collapse-stub>
        <b-button-group-stub tag="div" ariarole="group" id="submit-group" class="float-right">
          <b-button-stub target="_self" to="/db" event="click" routertag="a" variant="secondary" type="button" tag="button">Reset</b-button-stub>
          <b-button-stub target="_self" to="[object Object]" event="click" routertag="a" variant="primary" type="submit" tag="button" id="submit-search">Search</b-button-stub>
        </b-button-group-stub>
      </b-form-stub>
    </div>
    <div id="leads" infinite-scroll-disabled="disable_loading" infinite-scroll-distance="10" class="row justify-content-center">
      <lead-stub id="1420" headerlink="true"></lead-stub>
    </div>
    <!---->
  </div>
</login-required-stub>
`;
